### PR TITLE
[BEAM-2857] Implementing WriteToFiles transform for fileio (Python)

### PR DESCRIPTION
**Pull request by pabloem**
r: @chamikaramj

### CODE REVIEW

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Passing postcommit: https://builds.apache.org/job/beam_PostCommit_Python_Verify_PR/593/
PAssing py3 postcommit: https://builds.apache.org/job/beam_PostCommit_Python3_Verify_PR/230/

**Comment by udim** [comment]
still reviewing...

**Comment by pabloem** [comment]
ugh... I'm so sorry Udi. The design doc for this is here: https://s.apache.org/fileio-beam-python
it's light on details.. but i guess... it would be helpful : /

File: sdks/python/apache_beam/transforms/core.py [hunk 1]
```diff
@@ -1982,7 +1983,11 @@ def process(self, element, timestamp=DoFn.TimestampParam,
...
 
-  def __init__(self, windowfn, **kwargs):
+  def __init__(self,
+               windowfn,
+               trigger=None,
+               accumulation_mode=None,
+               timestamp_combiner=None):
     """Initializes a WindowInto transform.
 
     Args:
```
Comment (udim) [inline sdks/python/apache_beam/transforms/core.py]: Could you add the allowable types for `windowfn` (Windowing and WindowFn?).

File: sdks/python/apache_beam/testing/util.py [hunk 2]
```diff
@@ -141,6 +145,21 @@ def _equal(actual):
   return _equal
 
 
+def matches_all(expected):
```
Comment (udim) [inline sdks/python/apache_beam/testing/util.py]: Is this different from `contains_in_any_order` above?

Comment (pabloem) [reply sdks/python/apache_beam/testing/util.py]: The `contains_in_any_order` seems to require a set of hashable items. It simply does `collections.Counter()` on the iterable, and compares both counters. `matches_all`  takes in an iterable of elements or matchers, and matches them 1:1 properly.

File: sdks/python/apache_beam/io/fileio.py [hunk 3]
```diff
@@ -169,3 +179,476 @@ def __init__(self, compression=None, skip_directories=True):
...
+    if not self._temp_directory:
+      temp_location = (
+          p.options.view_as(GoogleCloudOptions).temp_location
+          or self.path.get())
+      dir_uid = str(uuid.uuid4())
+      self._temp_directory = StaticValueProvider(
+          str,
+          filesystems.FileSystems.join(temp_location,
+                                       '.temp%s' % dir_uid))
+      logging.info('Added temporary directory %s', self._temp_directory.get())
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/python/apache_beam/io/fileio.py]: No. The transform may be run in streaming, so I don't think there's a good moment in the pipeline execution to perform a cleanup of the temp directory.

**Comment by pabloem** [comment]
Hi Udi. I've addressed some of your comments. Some things pending to finalize:

- How do we clean up temporary files?
- In general, how do we deal with failures?

Some things that are worth looking at:

- I added documentation at the top of the file. LMK if that clarifies some things.
- Default sink is now a text sink that encodes strings to UTF8 encoding.

File: sdks/python/apache_beam/testing/util.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/testing/util.py]: gotcha

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
I've updated the cleanup code. Any error other than `BeamIOError` should be a retry. Also, I believe it should be okay to only clean up all of the files, and not the directory (as we'd have to check for batch vs streaming).
Thoughts?

**Comment by udim** [comment]
From what I understand deletion of temporary files is best effort. There is currently no mechanism to clean up old temporary files left over due to an error. Perhaps we could, in the future, add a TTL or some other way to tell if a temporary file can be safely deleted.

File: sdks/python/apache_beam/io/fileio.py [hunk 4]
```diff
@@ -554,6 +561,18 @@ def process(self, element):
                        r.pane,
                        destination)
 
+    logging.info('Cautiously removing temporary files for destination %s',
+                 destination)
+    self._remove_temporary_files([r.file_name for r in file_results])
```
Comment (udim) [inline sdks/python/apache_beam/io/fileio.py]: There's an implicit `return None` at the end of this method. Are `None`s handled correctly?

**Comment by udim** [comment]
Looks good, except for the new fileio.py comment I left above.

File: sdks/python/apache_beam/io/fileio.py (hunk 4 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/fileio.py]: I believe this method will return a generator due to the above yield keyword, so this does not return None.

**Comment by pabloem** [comment]
> From what I understand deletion of temporary files is best effort. There is currently no mechanism to clean up old temporary files left over due to an error. Perhaps we could, in the future, add a TTL or some other way to tell if a temporary file can be safely deleted.

Do you mean when the pipeline fails? (e.g. for batch?)

You're right about that, that if the pipeline fails, we won't be able to clean up the bundles that did not run. But how could we know this ahead of time? : /

Currently, all files will be cleaned up for any bundle that succeeds. If the bundle does not succeed, then we should not clean up the files.

**Comment by pabloem** [comment]
Run Python_PVR_Flink PreCommit

**Comment by udim** [comment]
Yes, it applies to cases where the pipeline fails, but could also be in cases where it is cancelled or updated (does the temp directory change in that case?).

We could format temp directory names to have the job ID in them, so it's possible to check if the job is still running and delete temp files if it isn't. This process would not be run as part of the pipeline.
Regarding temp directory deletion, I believe that for batch jobs they should be removed at the end (because it's better to clean up if we can). For streaming jobs, there's no hook for cleanup during cancellation and updates AFAIK.

Regarding failed bundles, assuming that the pipeline has not failed, it should be possible to scan for and delete any remaining temporary files.

**Comment by udim** [comment]
cc: @chamikaramj as I'm OOO next week.

**Comment by pabloem** [comment]
@chamikaramj

File: sdks/python/apache_beam/io/fileio.py [hunk 5]
```diff
@@ -502,12 +502,13 @@ def expand(self, pcoll):
...
-                            self.path, self.file_naming_fn)))
+                            self.path, self.file_naming_fn,
+                            self._temp_directory)))
 
     return file_results
 
 
-def _create_writer(base_path, file_name):
+def _create_writer(base_path, writer_key):
   # TODO(pabloem) Is this a good place to create a temporary directory?
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio.py]: Please create a JIRA or remove this.

File: sdks/python/apache_beam/io/fileio.py [hunk 6]
```diff
@@ -562,16 +569,21 @@ def process(self, element):
...
+  def _remove_temporary_files(self, writer_key):
     try:
-      filesystems.FileSystems.delete(files)
+      prefix = filesystems.FileSystems.join(
+          self.temporary_directory.get(), str(abs(hash(writer_key))))
+      match_result = filesystems.FileSystems.match(['%s*' % prefix])
+      orphaned_files = [m.path for m in match_result[0].metadata_list]
+
+      logging.debug('Deleting orphaned files: %s', orphaned_files)
+      filesystems.FileSystems.delete(orphaned_files)
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio.py]: Shouldn't we be ignoring missing source files to handle workitem retries ?

File: sdks/python/apache_beam/io/fileio.py [hunk 7]
```diff
@@ -170,3 +237,487 @@ def __init__(self, compression=None, skip_directories=True):
...
+              'compression': ''}
+    if total_shards is not None and shard_index is not None:
+      kwargs['shard'] = int(shard_index)
+      kwargs['total_shards'] = int(total_shards)
+
+    if window != GlobalWindow():
+      kwargs['start'] = window.start.to_utc_datetime().isoformat()
+      kwargs['end'] = window.end.to_utc_datetime().isoformat()
+
+    # TODO(pabloem): Add support for PaneInfo
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio.py]: Create a JIRA and mention here ?

File: sdks/python/apache_beam/io/fileio.py [hunk 8]
```diff
@@ -23,17 +23,83 @@
...
+================
+
+The transforms in this file include ``WriteToFiles``, which allows you to write
+a ``beam.PCollection`` to files, and gives you many options to customize how to
+do this.
+
+File Naming
+-----------
+One of the parameters received by ``WriteToFiles`` is a function specifying how
+to name the files that are written. This is a function that takes in the
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio.py]: Also document that this is supported for streaming and how it is supported.

File: sdks/python/apache_beam/io/fileio.py (hunk 6 above)
Comment (udim) [reply sdks/python/apache_beam/io/fileio.py]: I agree that we should ignore errors here. Note that `orphaned_files` are taken from a recent `filesystems.FileSystems.match()` call, so it doesn't make a difference if this is a retry.

File: sdks/python/apache_beam/io/fileio.py (hunk 5 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/fileio.py]: Done.

File: sdks/python/apache_beam/io/fileio.py (hunk 6 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/fileio.py]: Yes. All we're doing is logging the failures (in debug level) to be aware of them.

File: sdks/python/apache_beam/io/fileio.py (hunk 7 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/fileio.py]: Done

File: sdks/python/apache_beam/io/fileio.py (hunk 8 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/fileio.py]: Done.

**Comment by chamikaramj** [comment]
LGTM. Thanks.

Please fix-up and self-merge after tests pass.

**Comment by pabloem** [comment]
Run Python_PVR_Flink PreCommit

**Comment by chamikaramj** [comment]
Also, please fix the conflict.

**Comment by pabloem** [comment]
Run Portable_Python PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by yupbank** [comment]
Hey, So following the doc for which `AvroSink`  clearly doesn't exist and the `beam.io.avroio._AvroSink` would require much more info. For the context, i'm looking into ways to do `WriteToParquet` from a pubsub input. and `WriteToFiles` provides the closest thing. But the document is a bit confusing.

**Pull request merged**